### PR TITLE
add design document for "named-array"

### PR DESCRIPTION
**Pull request by andersy005**
This PR contains the design document for "named-array" a proposed new package aimed at providing a lightweight, efficient array structure with named dimensions. the main objective of named-array  is to provide a more accessible version of Xarray's Variable class without the heavy dependencies like Pandas.

~~## proposed name~~

~~The "Xarray-lite" name is a placeholder for the time being. we have considered two main candidates for the package's name: "narray" and "yarray." feedback on the proposed names or other suggestions is welcome.~~

Related issues:

  - https://github.com/pydata/xarray/issues/5081
  - https://github.com/pydata/xarray/issues/3981

### CODE REVIEW

File: design_notes/xarray_lite_design_doc.md [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+
+### Appendix: Implementation Details
+
+* Merge in VariableArithmetic's parent classes: AbstractArray, NdimSizeLenMixin with the new data structure..
+
+```python
+class VariableArithmetic(
+ ImplementsArrayReduce,
+ IncludeReduceMethods,
+ IncludeCumMethods,
```
Comment (andersy005) [inline design_notes/xarray_lite_design_doc.md:313]: could be removed if we generate the methods using `generate_aggregations.py`

File: design_notes/xarray_lite_design_doc.md [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+
+We propose the creation of a standalone Python package, "Xarray-lite" (actual name TBD). This package is envisioned to be a version of the "xarray.Variable" data structure, cleanly separated from the heavier dependencies of Xarray. Xarray-lite will provide a lightweight, user-friendly array-like data structure with named dimensions, facilitating convenient indexing and broadcasting. The package will use existing scientific Python community standards such as established array protocols and the new [Python array API standard](https://data-apis.org/array-api/latest), allowing users to wrap multiple duck-array objects, including, but not limited to, NumPy, Dask, Sparse, Pint, CuPy, and Pytorch.
```
Comment (andersy005) [inline design_notes/xarray_lite_design_doc.md:17]: we just decided to call the package "named-array" and to rename `Variable` -> `NamedArray`.

File: design_notes/xarray_lite_design_doc.md [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+        * Potentially implement this by introducing a new compliant wrapper object?
+    * Delete the `NON_NUMPY_SUPPORTED_ARRAY_TYPES` variable which special-cases ExplicitlyIndexed and `pd.Index.`
+        * `ExplicitlyIndexed` class and subclasses should provide `.oindex` and `.vindex` for indexing by `Variable.__getitem__.`: `oindex` and `vindex` were proposed in [NEP21](https://numpy.org/neps/nep-0021-advanced-indexing.html), but have not been implemented yet
+        * Delete the ExplicitIndexer objects (`BasicIndexer`, `VectorizedIndexer`, `OuterIndexer`)
+        * Remove explicit support for `pd.Index`. When provided with a `pd.Index` object, Variable will coerce to an array using `np.array(pd.Index)`. For Xarray's purposes, Xarray can use `as_variable` to explicitly wrap these in PandasIndexingAdapter and pass them to `Variable.__init__`.
+3. Define a minimal variable interface that the rest of Xarray can use:
+    1. `dims`: tuple of dimension names
+    2. `data`: numpy/dask/duck arrays`
+    3. `attrs``: dictionary of attributes
+    4. <del>Encoding: dict</del> ([GH](https://github.com/pydata/xarray/issues/6323))
```
Comment (dcherian) [inline design_notes/xarray_lite_design_doc.md:81]:
```suggestion
```

File: design_notes/named_array_design_doc.md [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+
+2. **Support for I/O**: named-array will not bundle file reading functions. Instead users will be expected to handle I/O and then wrap those arrays with the new named-array data structure.
+
+## Backward Compatibility
+
+The creation of named-array is intended to separate the `xarray.Variable` from Xarray into a standalone package. This allows it to be used independently, without the need for Xarray's dependencies, like Pandas. This separation has implications for backward compatibility.
+
+Since the new named-array is envisioned to contain the core features of Xarray's variable, existing code using Variable from Xarray should be able to switch to named-array with minimal changes. However, there are several potential issues related to backward compatibility:
+
+* **API Changes**: as the Variable is decoupled from Xarray and moved into named-array, some changes to the API may be necessary. These changes might include differences in function signature, etc. These changes could breaking existing code that relies on the current API and associated utility functions (e.g. `as_variable()`)
```
Comment (andersy005) [inline design_notes/named_array_design_doc.md:41]:
```suggestion
* **API Changes**: as the Variable is decoupled from Xarray and moved into named-array, some changes to the API may be necessary. These changes might include differences in function signature, etc. These changes could break existing code that relies on the current API and associated utility functions (e.g. `as_variable()`)
```

**Comment by dcherian** [comment]
I'm merging so we can advertise in `whats-new.rst`.

Thanks @andersy005 !

**Pull request merged**